### PR TITLE
streamingccl: enable snappy compression in transit

### PR DESCRIPTION
**Pull request by dt**
Release note: none.
Epic: none.

### CODE REVIEW

**Review by stevendanna — COMMENTED** [review]
Nice.

I'll let @msbutler review. 

I suppose we could do something like have the source send back whether it supports compression so that the destination knows whether it can successfully request it if we felt like making this on-by-default upgrade compatible for backporting. But I'm also happy to just not backport this or backport it default-off and then say that a user has to turn it on.

**Comment by dt** [comment]
>  backport it default-off 
yeah, this is why I made it a setting.

> have the source send back whether it supports compression 
I thought about this but it seems like more surface area/complexity than we want to keep long-term, particularly with the cloud-first nature of .0 and PCR's current availability.

I initially, on the decode path if decompress failed, tried to just un-marshal the "compressed" bytes in case we had a new consumer requesting compression from an old producer that just replied with uncompressed. I backed that out based on above thinking that we don't want long-term upgrade-specific cruft here, but could bring it back for a backport if we really want a better story than "toggle the setting".

**Review by msbutler — APPROVED** [review]
On the upgrade path: i'm in favor of minimizing upgrade cruft in the code, as I doubt anyone will actually need it. 23.2 or 24.1.0 c2c users can spin down and spin up a new job, as we've discussed.

**Comment by dt** [comment]
@stevendanna I think I and @msbutler are leaning towards just backporting this as-is, default on, and telling anyone upgrading from earlier versions to pause pcr or disable the setting during the upgrade -- or just power through the upgrade errors and all and catch up after you're on the new binary. Thoughts?

**Comment by dt** [comment]
@stevendanna I made it backwards compatible; the added complexity is only in the err!=nil branch so I decided I didn't mind.

File: pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go [hunk 1]
```diff
@@ -108,6 +108,13 @@ var ingestSplitEvent = settings.RegisterBoolSetting(
 	false,
 )
 
+var compress = settings.RegisterBoolSetting(
+	settings.SystemOnly,
+	"physical_replication.consumer.stream_compression.enabled",
+	"enables requesting a compressed stream from the producer when resumed",
+	true,
```
[the remark this answers is not in the document]
Comment (dt) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:115]: eh, I'd just as soon always have it on.

[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:115]: agreed, now that I have added context for our release strategy.

**Comment by dt** [comment]
TFTR!

bors r=msbutler

**Pull request merged**